### PR TITLE
modernize packaging

### PR DESCRIPTION
**Pull request by ocefpaf**
Use setuptools_scm instead of versioneer and pyproject.toml instead of setup.*.

Only review this one after #136 is sorted out.

### CODE REVIEW

File: lib/eofs/tests/test_coding_standards.py [hunk 1]
```diff
@@ -1,33 +0,0 @@
-"""Test coding standards compliance."""
```
Comment (ocefpaf) [inline lib/eofs/tests/test_coding_standards.py:1]: @ajdawson I replaced this test with pre-commits. We can add more checks there if you want but I would do that in another PR to avoid bloating this one more than it already is.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+where = ["lib"]
+[tool.setuptools.package-data]
+eofs_examples = [
+  "example_data/*",
+]
+eofs_tests = ["data/*"]
+[tool.setuptools.dynamic]
+readme = {file = "README.md", content-type = "text/markdown"}
+
+[tool.setuptools_scm]
```
Comment (ocefpaf) [inline pyproject.toml:55]: Using setuptools_scm instead of versioneer b/c that one is better maintained and "blessed" by the Python packaging community.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,75 @@
+[build-system]
```
Comment (ocefpaf) [inline pyproject.toml:1]: This replaces both setup.py and setup.cfg with "modern" python packaging practices.

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  pull_request:
+  push:
+    branches: [main]
+
+jobs:
+  run:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.9", "3.10"]
```
Comment (ocefpaf) [inline .github/workflows/tests.yml:13]: To support all the extras on Windows and Python 3.11 we would need to drop `cdms2`.
xref.: https://github.com/conda-forge/cdms2-feedstock/issues/80#issuecomment-1309980939 and https://github.com/conda-forge/cdtime-feedstock/pull/52#issuecomment-1302342070

**Review by ocefpaf — COMMENTED** [review]
@ajdawson let me know if you need help navigating this PR. Sorry I ended up bloating it but this will get the CIs back online/

**Comment by ocefpaf** [comment]
Twine check is failing with:

```
Checking eofs-0.1.dev1+g23a45fe-py3-none-any.whl: FAILED
ERROR    `long_description` has syntax errors in markup and would not be        
         rendered on PyPI.                                                      
         No content rendered from RST source.                                   
WARNING  `long_description_content_type` missing. defaulting to `text/x-rst`.   
Checking eofs-0.1.dev1+g23a45fe.tar.gz: PASSED with warnings
WARNING  `long_description_content_type` missing. defaulting to `text/x-rst`.   
WARNING  `long_description` missing.                                            
Error: Process completed with exit code 1.
```

I can reproduce that locally and I see that it still packages the deleted `setup.cfg`, no idea why/how that id happening :-/

File: .github/workflows/tests.yml (hunk 4 above)
Comment (ajdawson) [reply .github/workflows/tests.yml:13]: Maybe the time is coming to remove the cdms2 interface in this module...

Comment (ajdawson) [reply .github/workflows/tests.yml:13]: But we can leave it like this for now and deal with that separately.

File: lib/eofs/tests/test_coding_standards.py (hunk 1 above)
Comment (ajdawson) [reply lib/eofs/tests/test_coding_standards.py:1]: This is fine I think

**Comment by ajdawson** [comment]
Ok, so again I won't pretend that I understand every part of this, but I think this is going in the right direction so I will be happy to merge it. Thanks for the hard work (and it does look like quite a lot of work!). I'm not sure about the pypi secrets, you'll have to fill me in on what needs to be done there.

**Comment by ajdawson** [comment]
Sorry that took so long @ocefpaf, was locked out of PyPI account when I went to resolve this back in June, and then just forgot about it. I think this should all be set up properly now. Thanks for the effort!
